### ISSUE TITLE
Unable to identify the ASN of a network when connected to a low bandwidth network

### ISSUE THREAD

**Issue by chinmayisk** [ooni/probe-android#224]
## Expected Behavior
The ooni probe is able to identify the ASN of a network 

## Actual Behavior
The ooni probe on android as well as iOS is unable to identify the ASN of a network 

## Steps to Reproduce the Problem

  1. Run a website test / performance test on low bandwidth network 
  2. The test runs 
  3. Application records AS0

## Frequency 
Rare but was reproducable by me on different devices

## Specifications

  - Version: 2.0.3 
  - Platform: Android

  - Version: 2.0.6
  - Platform: iOS
]
![Screenshot_android](https://user-images.githubusercontent.com/1808797/58020774-caa07d80-7b26-11e9-9eef-b9027d0d313f.png)

![screenshot_iOS](https://user-images.githubusercontent.com/1808797/58020593-2f0f0d00-7b26-11e9-952a-5cf4b32d37ce.png)

**Comment by bassosimone** [comment]
Thanks for reaching out! I believe this bug is related to the fact that we do not retry the geolookup attempt. The new code that @hellais has written for the command line OONI is better in this respect and I am currently trying to understand how to expose this Go code to mobile apps.

**Comment by chinmayisk** [comment]
Thanks i might have an related feature.. shall i add that ?

**Comment by bassosimone** [comment]
Yes, please! If it’s deemed its own issue we can always create a new issue later :)

**Comment by xhdix** [comment]
The received xml file was correct at the time the ASN0 was displayed and now that it displays the country code incorrectly. (At both times, it was not a constant mistake)

![image](https://user-images.githubusercontent.com/12384263/58591013-0b5a6e00-826e-11e9-8299-5ee8669da9f0.png)

**Comment by lorenzoPrimi** [comment]
@bassosimone should this issue be moved in the mk repo?

**Comment by bassosimone** [comment]
@xhdix the issue is that we only use the Ubuntu XML to get the probe IP. Then we use MaxMind databases to map to the ASN, network name, and country code. Yet, our Android process for using a new version of such databases should be improved. See https://github.com/measurement-kit/measurement-kit/issues/1858.

@lorenzoPrimi Moving it to MK would most likely causing the issue to be closed. The new Go engine does better with respect to the issue reported here (and, as a matter of fact, I did another pass of extra robustness checks, i.e. exponential backoff, _after_ this issue was opened). These changes will start benefiting the Desktop version of OONI first and will benefit also the mobile version when we'll use the Go code in mobile apps). Because of this, I don't want to spend time fixing the issue in MK because that is valuable time I am not using to making the Go code available to mobile. That's why I'd close the issue as wontfix if moved to MK. I'd say we keep the issue here until it's fixed for Android.

**Comment by hellais** [comment]
@bassosimone is there a probe-engine issue that we can link to here so we remember to close that issue when it's fixed?

**Comment by bassosimone** [comment]
@hellais I think there are two issues here. One of them is certain: we have shipped releases of OONI with ageing ASN databases. The other is less certain: we assume that the low bandwidth network has caused the IP lookup to fail, but we are not sure. Assuming that the latter is the cause of the original issue here, then the actions we have taken in probe-engine/netx are: (1) the DNS lookup is retried either by Go code or by us, depending on who's doing the lookup; (2) fallback to other DNS mechanisms is available; `connect` and TLS handshake timeouts have been raised to 30 s.
